src/Catalyst.KBucket/KBucket.cs

```diff
@@ -0,0 +1,528 @@
+            return _Add(contact, out ping);
+        }
+
+        /// <summary>
+        ///   Splits the node, redistributes contacts to the new nodes, and marks the
+        ///   node that was split as an inner node of the binary tree of nodes by
+        ///   setting this.root.contacts = null
+        /// </summary>
+        private void _Split(Bucket<T> node, int bitIndex)
+        {
+            node.Left = new Bucket<T>();
+            node.Right = new Bucket<T>();
+
+            // redistribute existing contacts amongst the two newly created nodes
+            foreach (var contact in node.Contacts)
+            {
+                _DetermineNode(node, contact.Id, bitIndex)
+                   .Contacts.Add(contact);
+            }
+
+            node.Contacts = null; // mark as inner tree node
+
+            // don't split the "far away" node
+            // we check where the local node would end up and mark the other one as
+            // "dontSplit" (i.e. "far away")
+            var detNode = _DetermineNode(node, LocalContactId, bitIndex);
+            var otherNode = node.Left == detNode ? node.Right : node.Left;
+            otherNode.DontSplit = true;
+        }
+
+        /// <summary>
+        ///   Updates the contact selected by the arbiter.
+        /// </summary>
+        /// <remarks>
+        ///   If the selection is our old contact and the candidate is some new contact
+        ///   then the new contact is abandoned (not added).
+        ///
+        ///   If the selection is our old contact and the candidate is our old contact
+        ///   then we are refreshing the contact and it is marked as most recently
+        ///   contacted(by being moved to the right/end of the bucket array).
+        ///   
+        ///   If the selection is our new contact, the old contact is removed and the new
+        ///   contact is marked as most recently contacted.
+        /// </remarks>
+        private void _Update(Bucket<T> node, int index, T contact)
+        {
+            var incumbent = node.Contacts[index];
+            var selection = Arbiter(incumbent, contact);
+
+            // if the selection is our old contact and the candidate is some new
+            // contact, then there is nothing to do
+            if (selection == incumbent && incumbent != contact)
+            {
+                return;
+            }
+            
+            node.Contacts.RemoveAt(index);
+            node.Contacts.Add(selection);
+        }
+
+        /// <summary>
+        ///   Determines whether the id at the bitIndex is 0 or 1.
+        /// </summary>
+        /// <returns>
+        ///   Left leaf if `id` at `bitIndex` is 0, right leaf otherwise
+        /// </returns>
+        /// <remarks>
+        ///   This is an internal method.  It should not be directly called.
+        ///   It is only public for unit testing.
+        /// </remarks>
+        public Bucket<T> _DetermineNode(Bucket<T> node, byte[] id, int bitIndex)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Rename method '_DetermineNode' to match pascal case naming rules, trim underscores from the name.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)